uaclient/files.py

```diff
@@ -0,0 +1,344 @@
+import copy
+import json
+import logging
+import os
+from datetime import datetime
+from typing import Dict, Optional
+
+from uaclient import (
+    defaults,
+    event_logger,
+    exceptions,
+    messages,
+    schemas,
+    util,
+)
+from uaclient.machine_token_data import PublicMachineTokenData
+
+event = event_logger.get_event_logger()
+LOG = logging.getLogger(__name__)
+
+
+class UAFile:
+    def __init__(
+        self,
+        name: str,
+        root_directory: str = defaults.DEFAULT_DATA_DIR,
+        private: bool = True,
+    ):
+        self._root_directory = root_directory
+        self._file_name = name
+        self._is_private = private
+        self._path = None
+
+    @property
+    def path(self) -> str:
+        if not self._path:
+            self._path = os.path.join(
+                self._root_directory, self._file_name
+            )  # type: ignore
+        return self._path  # type: ignore
+
+    @property
+    def is_private(self) -> bool:
+        return self._is_private
+
+    def write(self, content):
+        file_mode = (
+            defaults.ROOT_READABLE_MODE
+            if self.is_private
+            else defaults.WORLD_READABLE_MODE
+        )
+        self.create_dir()
+        if not isinstance(content, str):
+            content = json.dumps(content, cls=util.DatetimeAwareJSONEncoder)
+        util.write_file(self.path, content, file_mode)
+
+    def read(self) -> Optional[str]:
+        content = None
+        try:
+            content = util.load_file(self.path)
+        except FileNotFoundError:
+            if not os.path.exists(self.path):
+                LOG.debug("File does not exist: {}".format(self.path))
+        return content
+
+    def delete(self):
+        util.remove_file(self.path)
+
+    def create_dir(self):
+        file_dir = os.path.dirname(self.path)
+        if not os.path.exists(file_dir):
```

Comment:
We already know the directory right?
```suggestion
        if not os.path.exists(self._root_directory):
```

Reply:
True.